### PR TITLE
feat: Add time-based cap

### PR DESCRIPTION
**Pull request by gitferry**
Closes: #96.

This PR added time-based cap to the global parameter and implemented the [algorithm](https://babylon-chain.atlassian.net/wiki/spaces/BABYLON/pages/297926664/Handle+time-based+staking+cap) to determine overflow staking transactions.

### CODE REVIEW

File: types/params.go [hunk 1]
```diff
@@ -43,3 +44,19 @@ func (pv *ParamsVersions) GetParamsForBTCHeight(btcHeight int32) (*GlobalParams,
 	}
 	return nil, fmt.Errorf("no version corresponds to the provided BTC height")
 }
+
+func (gp *GlobalParams) IsTimeBasedCap() (bool, error) {
+	if gp.StakingCap != 0 {
+		if gp.CapHeight != 0 {
+			return false, fmt.Errorf("only either of staking cap and cap height can be set")
```
Comment (KonradStaniec) [inline types/params.go:51]: We should probably panic here, as we are essentially running with invalid config. The only things we can do with this error later is to panic eitherway.

Also I would avoid nested ifs, as those are hard to read. Something like:

```
if gp.StakingCap != 0 && gp.CapHeight != 0 {
panic("only either of staking cap and cap height can be set")
}

if gp.StakingCap == 0 && gp.CapHeight == 0 {
panic("either of staking cap and cap height must be set")
}

return gp.CapHeight != 0
```

conveys the intend in much streigtforward fashion

File: indexer/indexer.go [hunk 2]
```diff
@@ -1090,13 +1090,31 @@ func (si *StakingIndexer) validateStakingTx(params *types.GlobalParams, stakingD
...
-func (si *StakingIndexer) isOverflow(cap uint64) (bool, error) {
+func (si *StakingIndexer) isOverflow(height uint64, params *types.GlobalParams) (bool, error) {
+	isTimeBased, err := params.IsTimeBasedCap()
+	if err != nil {
+		return false, err
+	}
+
+	if isTimeBased {
+		if height < params.ActivationHeight {
+			return false, fmt.Errorf("the transaction height %d should not be lower than the param activation height: %d",
```
Comment (KonradStaniec) [inline indexer/indexer.go:1101]: this is probably panic ? From what  I saw in logic later at this point we must have params for correct height.

Also lets avoid nested `if` if possible.

File: params/params.go [hunk 3]
```diff
@@ -46,6 +45,30 @@ func parseConfirmationDepthValue(confirmationDepth uint64) (uint16, error) {
 	return uint16(confirmationDepth), nil
 }
 
+// either staking cap and cap height should be positive
+// if cap height is positive, then it should be greater
+// than the activation height
+func parseCap(stakingCap, capHeight, activationHeight uint64) (btcutil.Amount, uint64, error) {
+	if stakingCap != 0 {
```
Comment (KonradStaniec) [inline params/params.go:52]: lets remove those nested ifs.

File: params/params.go [hunk 4]
```diff
@@ -46,6 +45,30 @@ func parseConfirmationDepthValue(confirmationDepth uint64) (uint16, error) {
...
+
+		parsedStakingCap, err := parseBtcValue(stakingCap)
+		return parsedStakingCap, 0, err
+	}
+
+	if capHeight == 0 {
+		return 0, 0, fmt.Errorf("either of staking cap and cap height must be set")
+	}
+
+	if capHeight <= activationHeight {
```
Comment (KonradStaniec) [inline params/params.go:65]: I wonder wheter we can enforce that. Take the case:
1. we got params with `{activationHeight: 100, capHeight:300}`
2. now from height `300` if we want to correct `unbondingFee`, params will have `{activationHeight: 300, capHeight:300}`

This seems to be valid update, but this check will refuse to do it.

**Review by KonradStaniec — COMMENTED** [review]
some minor comments, mostly about panicking if we discvoer we haver invalid parameters during runtime, as we can't handle those errors in any meaningful way either way.

File: types/params.go (hunk 1 above)
Comment (gitferry) [reply types/params.go]: Nice point! Fixed in [fb31a70](https://github.com/babylonchain/staking-indexer/pull/112/commits/fb31a70d4fa1bba5e6e5d297a2d2daba1c3040f1)

File: params/params.go (hunk 4 above)
Comment (gitferry) [reply params/params.go]: Makes sense. Fixed in [fb31a70](https://github.com/babylonchain/staking-indexer/pull/112/commits/fb31a70d4fa1bba5e6e5d297a2d2daba1c3040f1)

**Review by KonradStaniec — APPROVED** [review]

**Pull request merged**